velox/type/Conversions.h

```diff
@@ -418,6 +455,58 @@ struct Converter<TypeKind::VARCHAR, void, TRUNCATE, LEGACY_CAST> {
   static std::string cast(const bool& val) {
     return val ? "true" : "false";
   }
+
+  /// Normalize the given floating-point standard notation string in place, by
+  /// appending '.0' if it has only the integer part but no fractional part. For
+  /// example, for the given string '12345', replace with '12345.0'.
+  static void normalizeStandardNotation(std::string& str) {
+    if (!FLAGS_experimental_enable_legacy_cast &&
+        str.find(".") == std::string::npos && isdigit(str[str.length() - 1])) {
+      str += ".0";
+    }
+  }
+
+  /// Normalize the given floating-point scientific notation string in place, by
+  /// removing the trailing 0s of the coefficient as well as the leading '+' and
+  /// 0s of the exponent. For example, for the given string '3.0000000E+005',
+  /// replace with '3.0E5'. For '-1.2340000E-010', replace with '-1.234E-10'.
+  static void normalizeScientificNotation(std::string& str) {
+    auto idxE = str.find('E');
+    VELOX_DCHECK_NE(
+        idxE,
+        std::string::npos,
+        "Expect a character 'E' in scientific notation.");
+
+    auto endCoef = idxE - 1;
+    while (endCoef >= 0 && str[endCoef] == '0') {
+      --endCoef;
+    }
+    VELOX_DCHECK_GT(endCoef, 0, "Expect the decimal point at index 1.");
```

Comment:
nit: maybe "Expect the decimal point at index 1." --> "Coefficient should not be all zero."?

Reply:
Revised.